Open arch (#115)
Kuadrant/architecture: rfcs/0000-open-arch.md

Initially, users would only be exposed to these changes through _metapolicies_, domain specific policies that expose a
higher-level abstraction to our existing policies.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## What's a *Metapolicy*?

A *Metapolicy* is a policy just like any other [Gateway API Policy][2], other than it only produces one or more "core
Kuadrant policies". A *Metapolicy* is managed by a `MetapolicyController` that will interact with the
`KuadrantController` to integrate seamlessly in the ecosystem.

Think of a `MetapolicyController` being a pure function that takes the custom *Metapolicy* and the Gateway API network
objects it targets as input, and outputs one or more Kuadrant Policies as a result.

## Implementing a custom *Metapolicy*

Fundamentally a *Metapolicy* isn't much different from a regular Kubernetes [Custom
Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/), and as such will be
managed by some controller. But unlike a regular [Kubernetes
Controller](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers),

> it also needs to reconcile when changes to properties of the Gateway API it uses are observed. For that reason, the

Reviewer: ```suggestion
it also needs to reconcile when it observes changes to properties of the Gateway API it uses. For that reason, the
```

Or is it bot the controller that observes the changes?

Reviewer: In that case maybe: 

```suggestion
it also needs to reconcile when changes to properties of the Gateway API it uses are observed by the `KuadrantController`. For that reason, the
```

controller for a custom *Metapolicy* registers itself with the `KuadrantController`.

> [!TODO]
> Some initial SDK usage example

The example below is a modified example from the 
[controller-runtime's own example](https://github.com/kubernetes-sigs/controller-runtime/blob/main/examples/builtins/controller.go#L39).

```go
func (r *reconcileMetapolicy) Reconcile(ctx context.Context, request reconcile.Request, kuadrant *KuadrantContext) (reconcile.Result, error) {
  // Fetch the MetaPolicy from the cache
  rs := &user.MetaPolicy{}
  err := r.client.Get(ctx, request.NamespacedName, rs)
  if errors.IsNotFound(err) {
    log.Error(nil, "Could not find MetaPolicy")
    return reconcile.Result{}, nil
  }

  if err != nil {
    return reconcile.Result{}, fmt.Errorf("could not fetch MetaPolicy: %+v", err)
  }

  // Set the label if it is missing
  if rs.Labels == nil {
    rs.Labels = map[string]string{}
  }
	
  // resolve the CEL expression using the `KuadrantContext`

>   label := kuadrant.evaluateExpression("kuadrant.gatewaysFor(self)[0].metadata.name")

Author: yes... where `gatewaysFor` would probably provide "overlays", as CEL calls them, to act upon certain types. I don't know actually that we can do it, in this case, for anything that has `targetRef`, a la structural typing... so, might have to be `gatewaysFor(self.targetRef)`, where that then resolves to a known-type for with there is an overlay.

  if rs.Labels["gateway"] == label {
    return reconcile.Result{}, nil
  }

  // Update the MetaPolicy
  rs.Labels["gateway"] = label
  err = r.client.Update(ctx, rs)
  if err != nil {
    return reconcile.Result{}, fmt.Errorf("could not write MetaPolicy: %+v", err)
  }

  return reconcile.Result{}, nil
}
```

> It on differs in three ways from the original example:

Reviewer 2: ```suggestion
It only differs in three ways from the original example:
```

1. The `Reconcile` signature takes an additional parameter, the `kuadrant *KuadrantContext`;
1. It acts upon an hypothetical `user.MetaPolicy`;
1. It uses the `evaluateExpression` to resolve the CEL expression `kuadrant.gatewaysFor(self)[0].metadata.name` instead
   of using a fixed string for the label's value.

> Because that expression is evaluated upon creation, the `KuadrantRuntime` will make sure to reconcile whenever that cel

Reviewer 2: ```suggestion
Because that expression is evaluated upon creation, the `KuadrantRuntime` will make sure to reconcile whenever that CEL
```

expression's evaluated result would change; in this particular example whenever the name of the first `Gateway` changes.

> [!NOTE]
> Below is the proposed inital integration point for `MetapolicyController`s. The idea is to keep the deployment model
> fairly open moving forward. As of now, this would be the deployment model for _our own_ *Metapolicies*

To deploy our custom reconciler, we need to let the Kuadrant operator know about it, by adding it to the `Kuadrant` CR:

```yaml
apiVersion: kuadrant.io/v1beta2
kind: Kuadrant
metadata:
  name: kuadrant-sample
spec:
  plugins:
    userMetaPolicy:
      deployment: embedded
```

> The `UserMetaPolicy` will need to be packaged with a custom Docker image containing the plugin. It'll be automatically

Author: The `Plugin` in this case is the extension point that the Kuadrant controller needs to know about. It so happens that this extension is a "metacontroller" that effectively cares about changes to its own resources, e.g. `UserMetaPolicy` CRs _and_ the changes to relevant Gateway API resources (i.e. the DAG).

Now there is no reason the `Plugin` couldn't be more or something entirely different than a "metacontroller". It simply formalizes the API that such a plugin talks with the Kuadrant controller. That, and "sandboxes" it, so that plugins failing don't affect each other.

> to deal with failures et al, so probably adding the necessity for more configuration options for these plugins

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Plugin architecture

### gRPC to controller

- [ ] Unix socket for "in-pod/embedded" plugins
- [ ] Declarative way to load plugins
- [ ] Provide a decorated [Controller Runtime](https://github.com/kubernetes-sigs/controller-runtime) Golang that
  provides the changes triggered by the DAG

### gRPC streams for eventing on the DAG

- [ ] Subscriptions infered from CEL and the current CR being reconcilied

### Typed DSL in CEL

> [!IMPORTANT]
> Provide both the `MetapolicyCR` & the `GwAPIContext` instances relative to a reconciliation
> The `GwAPIContext` would not only be responsible for resolving the [CEL expressions][3] but also deriving the
> necessary subscriptions needed to inform the controller when a reconciliation is needed because of changes to the
> Gateway API objects queried

- [ ] Support existing `context`s in the CEL interpreter's `Env`
- [ ] Find a way to express the `context`s present/set for each CEL

### Deployment

> [!IMPORTANT]

> > This would initially be deployed as their own processes running alongside the kuadrant controller. They'd communicate

Reviewer 2: When you say along side. Does that mean in the same pod and as part of the same deployment object?

Reviewer 2: One consideration may be limiting the resources that such a container can have. I think we would also want to understand to protect the kuadrant operator from a bad plugin crashing and potentially causing the pod to get restarted

Reviewer 2: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-restarts

> though unix sockets, over gRPC & gRPC streams (for the eventing).
>
> Later, we could support different deployment models, as all that'd be needed is being able to use gRPC essentially.

- [ ] Known location for plugins to live in; one per directory

> - [ ] Unix socket opened within the same location?

Reviewer 2: Possibly use a volume mount of type emptyDir. Th https://kubernetes.io/docs/concepts/storage/volumes/#emptydir

Author: I was, for now certainly, not looking into making this a "first class citizen" per se, so you'd probably docker compose the base image to add e.g. the `/plugins/myPlugin/myPlugin` executable to the kuadrant operator.

# Drawbacks
[drawbacks]: #drawbacks

> [!todo]
> Why should we *not* do this?

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> [!todo]
> - Why is this design the best in the space of possible designs?
> - What other designs have been considered and what is the rationale for not choosing them?
> - What is the impact of not doing this?

# Prior art
[prior-art]: #prior-art

> [!todo]
> Discuss prior art, both the good and the bad, in relation to this proposal.
> A few examples of what this can include are:
>

> > - Does another project have a similar feature?

Reviewer 3: Similar to ext_proc in envoy proxy? https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter
